java-engine/src/main/java/io/getunleash/engine/UnleashEngine.java

```diff
@@ -1,62 +1,86 @@
+        YggResponse<Void> response = read(yggdrasil.takeState(toggles), new TypeReference<>() {});
         if (!response.isValid()) {
             throw new YggdrasilInvalidInputException(toggles);
         }
     }
 
-    public boolean isEnabled(String name, Context context) throws YggdrasilInvalidInputException {
+    public Boolean isEnabled(String name, Context context) throws YggdrasilInvalidInputException, YggdrasilError {
         try {
-            String jsonContext = mapper.writeValueAsString(context);
-            IsEnabledResponse isEnabled = read(yggdrasil.checkEnabled(name, jsonContext), IsEnabledResponse.class);
-            return isEnabled.isEnabled();
+            String jsonContext = writer.writeValueAsString(context);
+            YggResponse<Boolean> isEnabled = read(yggdrasil.checkEnabled(name, jsonContext), new TypeReference<>() {});
+            return isEnabled.getValue();
         } catch (JsonProcessingException e) {
             throw new YggdrasilInvalidInputException(context);
         }
     }
 
-    public VariantResponse getVariant(String name, Context context) throws YggdrasilInvalidInputException {
+    public VariantDef getVariant(String name, Context context) throws YggdrasilInvalidInputException, YggdrasilError {
         try {
-            String jsonContext = mapper.writeValueAsString(context);
-            return read(yggdrasil.checkVariant(name, jsonContext), VariantResponse.class);
+            String jsonContext = writer.writeValueAsString(context);
+            YggResponse<VariantDef> response = read(yggdrasil.checkVariant(name, jsonContext), new TypeReference<>() {});
+            return response.getValue();
         } catch (JsonProcessingException e) {
             throw new YggdrasilInvalidInputException(context);
         }
     }
 
+    public void countToggle(String flagName, boolean enabled) {
+        this.yggdrasil.countToggle(flagName, enabled);
+    }
+
+    public void countVariant(String flagName, String variantName) {
+        this.yggdrasil.countVariant(flagName, variantName);
+    }
+
+    public MetricsBucket getMetrics() throws YggdrasilError {
+        YggResponse<MetricsBucket> response = read(yggdrasil.getMetrics(), new TypeReference<>() {});
+        return response.getValue();
+    }
+
     /**
      * Handle reading from a pointer into a String and mapping it to an object
      */
-    private <T> T read(Pointer pointer, Class<T> clazz) {
+    private <T> T read(Pointer pointer, TypeReference<T> typeReference) {
         String str = pointer.getString(0, UTF_8);
         yggdrasil.freeResponse(pointer);
         try {
             System.out.println(str); // TODO use a logging library. SLF4J?
-            return mapper.readValue(str, clazz);
-        } catch (JsonProcessingException e) {
-            throw new YggdrasilParseException(str, clazz, e);
+            return reader.forType(typeReference).readValue(str);
```

Comment:
Potential for optimization:
- reader.forType(typeReference)
could be in a Map for each one of the known types...